File: remote/src/main/java/com/kelsos/mbrc/interactors/library/ArtistAlbumInteractorImpl.java
package com.kelsos.mbrc.interactors.library;

import com.google.inject.Inject;
import com.kelsos.mbrc.domain.Album;
import com.kelsos.mbrc.mappers.AlbumMapper;
import com.kelsos.mbrc.repository.library.AlbumRepository;
import java.util.List;
import rx.Observable;

public class ArtistAlbumInteractorImpl implements ArtistAlbumInteractor {
  @Inject private AlbumRepository repository;

  @Override public Observable<List<Album>> getArtistAlbums(long artistId) {
    return repository.getAlbumsByArtist(artistId)
        .flatMap(albums -> Observable.defer(() -> Observable.just(AlbumMapper.mapArtistAlbums(albums))));
  }
}

File: app/src/main/kotlin/com/kelsos/mbrc/interactors/library/ArtistAlbumInteractorImpl.kt
package com.kelsos.mbrc.interactors.library

import com.kelsos.mbrc.domain.Album
import com.kelsos.mbrc.mappers.AlbumMapper
import com.kelsos.mbrc.repository.library.AlbumRepository
import rx.Observable
import rx.lang.kotlin.toSingletonObservable
import javax.inject.Inject

class ArtistAlbumInteractorImpl
@Inject constructor(private val repository: AlbumRepository): ArtistAlbumInteractor {

  override fun getArtistAlbums(artistId: Long): Observable<List<Album>> {
    return repository.getAlbumsByArtist(artistId)
        .flatMap {
          Observable.defer {
            AlbumMapper.mapArtistAlbums(it).toSingletonObservable()
          }
        }
  }
}
